R/effectsize.R

```diff
@@ -0,0 +1,153 @@
+#' # run a linear model relating the data in all voxels to Genotype
+#' vs <- mincLm(jacobians_fixed_2 ~ Sex, gf)
+#' effectsize <- mincEffectSize(vs)
+#' }
+#' @export
+vertexEffectSize <- function(buffer, predictors = NULL)
+{
+  #Nakagawa, S., Cuthill, I.C., 2007. Effect size, confidence interval and statistical significance: a practical guide for biologists. Biol. Rev. Camb. Philos. Soc. 82, 591–605. https://doi.org/10.1111/j.1469-185X.2007.00027.x
+  #Original unbiased corrector from paper replaced with
+  #unbiased corrector function J from https://en.wikipedia.org/wiki/Effect_size#Hedges'_g
+  #Watch out for exploding gamma function
+  J <- function(a) {
+    out <- tryCatch({
+      return (gamma(a / 2) / (sqrt(a / 2) * gamma((a - 1) / 2)))
+    },
+    error = function(cond) {
+      return(1)
+    },
+    warning = function(cond) {
+      return(1)
+    })
+    return(out)
+  }
+  # g_biased = t ( n1 + n2 ) / sqrt(n1*n2)*sqrt(df)
+  # g_unbiased = g_biased * J(n1 + n2 - 2)
+  # g_variance_unbiased = (n1 + n2) / (n1*n2) + g_unbiased**2 / (2 * (n1 + n2 - 2))
+  # n1 study group
+  # n2 reference group
+  # N = n1+n2
+  # NN = n1*n2
+  
+  originalMincAttrs <- mincAttributes(buffer)
+  stattype <- originalMincAttrs$`stat-type`
+  # Remove coefficients from buffer
+  if (!is.null(stattype)) {
+    for (nStat in 1:length(stattype)) {
+      if (stattype[nStat] == 'beta' ||
+          stattype[nStat] == 'R-squared' ||
+          stattype[nStat] == "logLik" || stattype[nStat] == "F") {
+        if (!exists('indicesToRemove')) {
+          indicesToRemove = nStat
+        }
+        else {
+          indicesToRemove = c(indicesToRemove, nStat)
+        }
+      }
+    }
+    if (exists('indicesToRemove')) {
+      updatedAttrs <- originalMincAttrs
+      updatedAttrs$`stat-type` <-
+        updatedAttrs$`stat-type`[-indicesToRemove]
+      updatedAttrs$dimnames[[2]] <-
+        updatedAttrs$dimnames[[2]][-indicesToRemove]
+      
+      buffer <- buffer[, -indicesToRemove]
+      buffer <- setMincAttributes(buffer, updatedAttrs)
+    }
+  }
+  
+  model_call <- attr(buffer, "call")
+  model_call[[1]] <- quote(model.matrix)
+  names(model_call)[names(model_call) == "formula"] <- ""
+  mod_mat <- eval(model_call)
+  conts <- attr(mod_mat, "contrasts")
+  if (is.null(conts))
+    stop("No categorical variables in model, cannot compute g* statistics")
+  if (!all(sapply(predictors, function(x) any(grepl(x, updatedAttrs$dimnames[[2]])))))
+    stop("Supplied predictors not found in model")
+  if (!all(conts == "contr.treatment"))
+      stop("Non-treatment factors present in model, effect size is not meaningful
+           for this configuration")
+
+  if (is.null(predictors)) {
+    cat_vars <-
+      names(conts)[vapply(conts, function(x)
+        all(x == "contr.treatment"), logical(1))]
+    predictors <- lapply(updatedAttrs$dimnames, function(x)
+      grep(paste(cat_vars, collapse = "|"), x, value = TRUE))[[2]]
```

Comment:
```suggestion
      grep(paste(cat_vars, collapse = "|"), x, value = TRUE))
```

Reply:
I believe I had that before and it didn't work.

Reply:
Ah, I see you moved it.

Reply:
Looks like you don't need the lapply anymore.

Reply:
Not sure what apply would work instead? "apply" needs a direction to go.

Reply:
And "vapply" needs a "FUN.VALUE" whatever that is.

Reply:
You don't need any. Grep is vectorized in the searched strings dimensions. Plug colnames...[[2]] into x

Reply:
Right...